File: mobile/ios/packages/town-crier-presentation/Sources/Features/ForceUpdate/ForceUpdateViewModel.swift
import Combine
import TownCrierDomain

/// ViewModel that checks whether the current app version meets the minimum
/// required version. If not, exposes state for a blocking update modal.
@MainActor
public final class ForceUpdateViewModel: ObservableObject {
    @Published public private(set) var requiresUpdate = false
    @Published public private(set) var isChecking = false

    private let versionConfigService: VersionConfigService
    private let appVersionProvider: AppVersionProvider

    public init(
        versionConfigService: VersionConfigService,
        appVersionProvider: AppVersionProvider
    ) {
        self.versionConfigService = versionConfigService
        self.appVersionProvider = appVersionProvider
    }

    public func checkVersion() async {
        isChecking = true
        defer {
            isChecking = false
        }

        guard let currentVersion = AppVersion(appVersionProvider.version) else {
            return
        }

        do {
            let minimumVersion = try await versionConfigService.fetchMinimumVersion()
            requiresUpdate = currentVersion < minimumVersion
        } catch {
            // On failure, allow the app to continue — don't block users
        }
    }
}

File: mobile/android/presentation/src/main/kotlin/uk/towncrierapp/presentation/features/forceupdate/ForceUpdateViewModel.kt
package uk.towncrierapp.presentation.features.forceupdate

import uk.towncrierapp.domain.auth.DomainError
import uk.towncrierapp.domain.versionconfig.AppVersion
import uk.towncrierapp.domain.versionconfig.VersionConfigService
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Blocking pre-login version gate: unauthenticated `GET /v1/version-config`
 * versus [currentVersion] (`BuildConfig.VERSION_NAME`, injected from `:app`
 * since `:presentation` cannot see another module's generated BuildConfig).
 * Re-checked on every auth-state transition (epic #770). A failed check
 * never blocks the user — port of iOS `ForceUpdateViewModel`.
 */
public class ForceUpdateViewModel(
    private val versionConfigService: VersionConfigService,
    private val currentVersion: String,
) : ViewModel() {
    private val _uiState = MutableStateFlow(ForceUpdateUiState())
    public val uiState: StateFlow<ForceUpdateUiState> = _uiState.asStateFlow()

    public fun checkVersion() {
        val runningVersion = AppVersion.parse(currentVersion) ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(isChecking = true) }
            try {
                val minimumVersion = versionConfigService.fetchMinimumVersion()
                _uiState.update { it.copy(isChecking = false, requiresUpdate = runningVersion < minimumVersion) }
            } catch (e: CancellationException) {
                throw e
            } catch (e: DomainError) {
                // Fail open — a version-config outage must never lock users out.
                _uiState.update { it.copy(isChecking = false) }
            }
        }
    }
}
